### ISSUE TITLE
Various issues with transparent materials

### ISSUE THREAD

**Issue by AlexECX** [alloystorm/dvvr#361]
**Describe the bug**
#### 1. Material that is transparent in PMX, but isn't loaded as transparent in DanceXR

This one is not too bad to workaround, but it would be nice if it worked without tweaks.

These glasses are set to opaque in PMX while still being transparent when rendered in the editor, but are opaque in DanceXR:

![image](https://github.com/user-attachments/assets/14edf97c-3b83-4511-a71a-1bb439252e00)

The current workaround is to set Materials->Transparent Materials to `On`, manually select the glasses, and set 
Transparent Materials->Options->Transparent: `Force Transparent`.

Zip with the problematic component:
[zipp.zip](https://github.com/user-attachments/files/16811106/zipp.zip)

Screenshots of transparency in the PMX editor:

![image](https://github.com/user-attachments/assets/aad7bb52-d356-4f62-8d5e-539f9d0b9562)
![image](https://github.com/user-attachments/assets/9f06d1c6-5d5c-4a77-9679-21c5a3fc7f2c)

#### 2. Material that is transparent in PMX and loaded as transparent in DanceXR, but turns opaque right after loading

Similar to the previous case, except in DanceXR the texture is transparent during the loading animation, but quickly turns opaque right after. This usually happens with materials meant to be hidden/shown using a morph that either changes the opacity or in this case moves it into place. These materials often come with empty space/texture around them that should be transparent.

https://github.com/user-attachments/assets/38ae05d6-3094-4ccf-8a8d-a661ffd9e2ee

Zip with the problematic component:
[stockings.zip](https://github.com/user-attachments/files/16821174/stockings.zip)

The current workaround is the same as for the previous case, force as transparent.

**Expected behavior**
The transparent part of the materials should be transparent by default

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Windows 11    Version	10.0.22631 Build 22631
 - DanceXR: Pro version 2024.8 (Steam build ID: 15381832) in desktop mode

**Comment by AlexECX** [comment]
Updated the stockings zip to add an additional file. I was preparing a third example, where the material's Alpha=0 in PMXE but DanceXR would still render it as opaque, when I noticed that it was using the same texture file with different alpha values and vertices to represent 2 different states of the material on the model. It makes me wonder if DanceXR is applying the render parameters of the first material and apply them to the second.

In the case of `#2` the stockings are first set as an opaque texture and nothing more (to be modifiable by a UV morph) but then referenced again with additional skin parameters (probably hide the blank parts of the texture and only show the foot fishnet)

![image](https://github.com/user-attachments/assets/1e39540a-129b-4d0a-b50f-d67f8299f02a)
![image](https://github.com/user-attachments/assets/96ebdb79-5660-4aa9-8a92-042af6622950)


In this `#3` case I'm working with, the texture is first referenced as opaque then as transparent (and with different vertices)

Zip with the problematic component:
[dual underwear.zip](https://github.com/user-attachments/files/16823101/dual.underwear.zip)


![image](https://github.com/user-attachments/assets/0f6595f6-0ec7-4d88-ac54-b3380c6595b4)
![image](https://github.com/user-attachments/assets/66992ff7-02f8-4766-af0d-cc40357ef2d4)

it actually has 2 problems:
- the regular underwear material (Alpha=1) has an outline that should be transparent but appears black until forced to transparent.
- the alternative underwear material (Alpha=0) does not have the previous problem, but it appears as an outline on the regular underwear and need to be set to Alpha=0 manually in DanceXR


https://github.com/user-attachments/assets/6434a603-7e50-4f0c-959c-dc1d212cb1c2

**Comment by AlexECX** [comment]
RIP deleted my comment by accident, this will be shorter.

TLDR material that use the same texture but with diff parameters is the problem maybe?

In the case of `#2` stockings, the texture is first referred to with no additional parameters (probably what is moved by the UV morph) and then referred to with additional parameters (probably to render only the foot fishnet):

![image](https://github.com/user-attachments/assets/9a0ad628-1139-407c-ac52-51d456a62ae9)
![image](https://github.com/user-attachments/assets/172b4c12-c4b7-4207-9c6b-9666740d3a01)


In a new case `#3` I'm working with 2 variation of the same clothing/material refer to the same texture, first with Alpha=1 (regular underwear) then Alpha=0 (the hidden variation).

![image](https://github.com/user-attachments/assets/1a36f468-9f6a-44e4-8c56-a4f581bbf02e)
![image](https://github.com/user-attachments/assets/512ed0cf-2647-45b3-b72a-4f4ef7f9ba68)


This new case has 2 problems:
- the regular underwear has an outline that should be transparent but is rendered black, workaround is to force opaque or transparent.
- the variation creates an outline on the regular underwear that can only be solved by setting the variation to force opaque (so it can still be used later) and manually set to Alpha=0


https://github.com/user-attachments/assets/593e2cb6-e23d-4367-9b1f-0fb538134f27

**Comment by alloystorm** [comment]
Thanks for the upload we'll take a look. 

For transparency, the PMX format doesn't really specify which texture is transparent, so we can only rely on the alpha value and the material name to "guess" if a material is suppose to be transparent. That's why when you set opaque to 1 it's not transparent at all. We believe MMD simply treat all materials as transparent, but we don't have that option because in unity transparent materials needs to be sorted properly to appear correct (MMD must have its own rendering technique that does some sort of auto sorting). 
So currently for some models you need to manually go to the material list to change their transparency mode if the model doesn't look right.

**Comment by AlexECX** [comment]
> For transparency, the PMX format doesn't really specify which texture is transparent, so we can only rely on the alpha value and the material name to "guess" if a material is suppose to be transparent. 

Yes and I think for `#1` having to manually change the sunglasses to an opacity lower than 1 is fair, I'm not even sure how PMXE does it internally. For `#3` tho the material **_is_** set to opaque 0, but my guess is since the texture has already been refered to by a different material that is set to opaque 1 it just sets both to opaque?

**Comment by AlexECX** [comment]
For `#3`, I tried:
1.  Inverting the order of the materials in the PMX. It did put the alternative underwear on top but still not transparent (even with opacity 0 in PMX)
2. Deleting the standard underwear material in the PMX. The alternative underwear is still visible with opacity 0, so it seems opacity is secondary to some other parameter.

**Comment by AlexECX** [comment]
Updated to version 2024.9

**Comment by AlexECX** [comment]
Updated to 2024.11 and added case `#3` to main post.

**Comment by alloystorm** [comment]
Not sure if you are still following this but the enhanced texture support in 2025.8 (with further fixes in 2025.9) should address most of  these and provide a lot more flexibility than before.

Closing this for now and feel free to open new issues if there are any.
